File: typescript/base/src/main/vsdk/toolkit/galerkin/processing/GatheringClusteredStrategy.ts
import { RendererConfiguration } from "../../material/RendererConfiguration";
import { GalerkinBasis } from "../GalerkinBasis";
import { GalerkinElement } from "../GalerkinElement";
import { GalerkinRadianceMethod } from "../GalerkinRadianceMethod";
import { GalerkinRole } from "../GalerkinRole";
import { GalerkinState } from "../GalerkinState";
import { Potential } from "../../render/Potential";
import { Scene } from "../../scene/Scene";
import { ElementFlags } from "../../environment/geometry/elements/ElementFlags";
import { Patch } from "../../environment/geometry/elements/Patch";
import { GatheringStrategy } from "./GatheringStrategy";
import { HierarchicalRefinementStrategy } from "./HierarchicalRefinementStrategy";
import { LinkingClusteredStrategy } from "./LinkingClusteredStrategy";

export class GatheringClusteredStrategy extends GatheringStrategy {
  private static updatePotential(cluster: GalerkinElement): number {
    if ((cluster.flags & ElementFlags.IS_CLUSTER_MASK) !== 0) {
      cluster.potential = 0.0;
      for (let i = 0; cluster.irregularSubElements !== null && i < cluster.irregularSubElements.length; i++) {
        if (!(cluster.irregularSubElements[i] instanceof GalerkinElement)) {
          continue;
        }
        const subCluster = cluster.irregularSubElements[i] as GalerkinElement;
        cluster.potential += subCluster.area * GatheringClusteredStrategy.updatePotential(subCluster);
      }
      cluster.potential /= cluster.area;
    }
    return cluster.potential;
  }

  private static updateClusterDirectPotential(element: GalerkinElement, potentialIncrement: number): void {
    if (element.regularSubElements !== null) {
      for (let i = 0; i < 4; i++) {
        if (element.regularSubElements[i] instanceof GalerkinElement) {
          GatheringClusteredStrategy.updateClusterDirectPotential(
            element.regularSubElements[i] as GalerkinElement,
            potentialIncrement,
          );
        }
      }
    }
    element.directPotential += potentialIncrement;
    element.potential += potentialIncrement;
  }

  public constructor() {
    super();
  }

  public override doGatheringIteration(scene: Scene, galerkinState: GalerkinState, renderOptions: RendererConfiguration): boolean {
    if (galerkinState.importanceDriven !== 0
      && (galerkinState.iterationNumber <= 1 || (scene.camera as NonNullable<Scene["camera"]>).changed !== 0)) {
      Potential.updateDirectPotential(scene, renderOptions);
      for (let i = 0; scene.patchList !== null && i < scene.patchList.length; i++) {
        const patch = scene.patchList[i];
        const top = patch.radianceData as GalerkinElement;
        const potentialIncrement = patch.directPotential - top.directPotential;
        GatheringClusteredStrategy.updateClusterDirectPotential(top, potentialIncrement);
      }
      GatheringClusteredStrategy.updatePotential(galerkinState.topCluster as GalerkinElement);
      (scene.camera as NonNullable<Scene["camera"]>).changed = 0;
    }

    process.stdout.write(`Galerkin (clustered) iteration ${galerkinState.iterationNumber}\n`);

    if (galerkinState.iterationNumber <= 1) {
      LinkingClusteredStrategy.createInitialLinks(
        galerkinState.topCluster as GalerkinElement,
        GalerkinRole.RECEIVER,
        galerkinState,
      );
    }

    const userErrorThreshold = galerkinState.relLinkErrorThreshold;
    HierarchicalRefinementStrategy.refineInteractions(scene, galerkinState.topCluster as GalerkinElement, galerkinState);
    galerkinState.relLinkErrorThreshold = userErrorThreshold;

    GalerkinBasis.pushPullRadiance(galerkinState.topCluster, galerkinState);

    if (galerkinState.importanceDriven !== 0) {
      GatheringStrategy.pushPullPotential(galerkinState.topCluster, 0.0);
    }

    galerkinState.ambientRadiance.clear();

    for (let i = 0; scene.patchList !== null && i < scene.patchList.length; i++) {
      GalerkinRadianceMethod.recomputePatchColor(scene.patchList[i]);
    }
    return false;
  }
}

File: cpp/base/src/main/vsdk/toolkit/galerkin/processing/GatheringClusteredStrategy.cpp
#include "java/util/ArrayList.txx"
#include "vsdk/toolkit/render/Potential.h"
#include "vsdk/toolkit/galerkin/GalerkinBasis.h"
#include "vsdk/toolkit/galerkin/GalerkinRadianceMethod.h"
#include "vsdk/toolkit/galerkin/processing/GatheringClusteredStrategy.h"
#include "vsdk/toolkit/galerkin/processing/HierarchicalRefinementStrategy.h"
#include "vsdk/toolkit/galerkin/processing/LinkingClusteredStrategy.h"

GatheringClusteredStrategy::GatheringClusteredStrategy() {
}

GatheringClusteredStrategy::~GatheringClusteredStrategy() {
}

/**
Updates the potential of the element after a change of the camera, and as such
a potential change in directly received potential
*/
void
GatheringClusteredStrategy::updateClusterDirectPotential(GalerkinElement *element, float potential_increment) {
    if ( element->regularSubElements != nullptr ) {
        for ( int i = 0; i < 4; i++ ) {
            GatheringClusteredStrategy::updateClusterDirectPotential(static_cast<GalerkinElement *>(element->regularSubElements[i]), potential_increment);
        }
    }
    element->directPotential += potential_increment;
    element->potential += potential_increment;
}

/**
Recomputes the potential of the cluster and its sub-clusters based on the
potential of the contained patches
*/
float
GatheringClusteredStrategy::updatePotential(GalerkinElement *cluster) {
    if ( cluster->flags & ElementFlags::IS_CLUSTER_MASK ) {
        cluster->potential = 0.0;
        for ( int i = 0; cluster->irregularSubElements != nullptr && i < cluster->irregularSubElements->size(); i++ ) {
            GalerkinElement *subCluster = static_cast<GalerkinElement *>(cluster->irregularSubElements->get(i));
            cluster->potential += subCluster->area * GatheringClusteredStrategy::updatePotential(subCluster);
        }
        cluster->potential /= cluster->area;
    }
    return cluster->potential;
}

/**
Note: clustering should not be turned off during the calculations
*/
bool
GatheringClusteredStrategy::doGatheringIteration(const Scene *scene, GalerkinState *galerkinState, RendererConfiguration *renderOptions) {
    if ( galerkinState->importanceDriven &&
        ( galerkinState->iterationNumber <= 1 || scene->camera->changed ) ) {
        Potential::updateDirectPotential(scene, renderOptions);
        for ( int i = 0; scene->patchList != nullptr && i < scene->patchList->size(); i++ ) {
            Patch *patch = scene->patchList->get(i);
            GalerkinElement *top = static_cast<GalerkinElement *>(patch->getRadianceData());
            float potentialIncrement = patch->getDirectPotential() - top->directPotential;
            GatheringClusteredStrategy::updateClusterDirectPotential(top, potentialIncrement);
        }
        GatheringClusteredStrategy::updatePotential(galerkinState->topCluster);
        scene->camera->changed = false;
    }

    java::System::out.printf("Galerkin (clustered) iteration %i\n", galerkinState->iterationNumber);

    // Initial linking stage is replaced by the creation of a self-link between
    // the whole scene and itself
    if ( galerkinState->iterationNumber <= 1 ) {
        LinkingClusteredStrategy::createInitialLinks(galerkinState->topCluster, GalerkinRole::RECEIVER, galerkinState);
    }

    double userErrorThreshold = galerkinState->relLinkErrorThreshold;

    // Refines and computes light transport over the refined links
    HierarchicalRefinementStrategy::refineInteractions(scene, galerkinState->topCluster, galerkinState);

    // TODO: This makes galerkinState non const. Check if this can be changed
    galerkinState->relLinkErrorThreshold = static_cast<float>(userErrorThreshold);

    // Push received radiance down the hierarchy to the leaf elements, where
    // it is multiplied with the reflectivity and the self-emitted radiance added,
    // and finally pulls back up for a consistent multi-resolution representation
    // of radiance over all levels
    GalerkinBasis::pushPullRadiance(galerkinState->topCluster, galerkinState);

    if ( galerkinState->importanceDriven ) {
        GatheringStrategy::pushPullPotential(galerkinState->topCluster, 0.0);
    }

    // No visualisation with ambient term for gathering radiosity algorithms
    galerkinState->ambientRadiance.clear();

    // Update the display colors of the patches
    for ( int i = 0; scene->patchList != nullptr && i < scene->patchList->size(); i++ ) {
        GalerkinRadianceMethod::recomputePatchColor(scene->patchList->get(i));
    }

    return false; // Never done
}
